File: 995/995.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/995.py

using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;

class Program
{
    const int LIMIT = 20000;
    const int PRIME_SEARCH_LIMIT = 2000000;

    static List<int> Sieve(int n)
    {
        var isPrime = new bool[n + 1];
        for (int i = 0; i <= n; i++) isPrime[i] = true;
        if (n >= 0) isPrime[0] = false;
        if (n >= 1) isPrime[1] = false;
        int r = (int)Math.Sqrt(n);
        for (int i = 2; i <= r; i++)
            if (isPrime[i])
                for (int j = i * i; j <= n; j += i) isPrime[j] = false;
        var result = new List<int>();
        for (int i = 0; i <= n; i++) if (isPrime[i]) result.Add(i);
        return result;
    }

    static readonly List<int> PRIMES = Sieve(PRIME_SEARCH_LIMIT);

    static List<(BigInteger, int)> FactorN(BigInteger n)
    {
        var outList = new List<(BigInteger, int)>();
        BigInteger t = n;
        foreach (int p in PRIMES)
        {
            BigInteger pb = new BigInteger(p);
            if (pb * pb > t) break;
            if (t % pb == 0)
            {
                int e = 0;
                while (t % pb == 0) { t /= pb; e++; }
                outList.Add((pb, e));
            }
        }
        if (t > BigInteger.One) outList.Add((t, 1));
        return outList;
    }

    static List<BigInteger> DivisorsFromFactorization(List<(BigInteger p, int e)> factors)
    {
        var divs = new List<BigInteger> { BigInteger.One };
        foreach (var (p, e) in factors)
        {
            var old = divs.ToList();
            divs = new List<BigInteger>();
            BigInteger power = BigInteger.One;
            for (int i = 0; i <= e; i++)
            {
                foreach (var d in old) divs.Add(d * power);
                power *= p;
            }
        }
        divs.Sort();
        return divs;
    }

    static BigInteger ModPow(BigInteger b, BigInteger e, BigInteger mod) => BigInteger.ModPow(b, e, mod);

    static BigInteger PrimitiveRoot(BigInteger p, List<BigInteger> primeFactorsOfPMinus1)
    {
        if (p == 2) return BigInteger.One;
        BigInteger m = p - 1;
        for (BigInteger g = 2; g < p; g++)
        {
            bool ok = true;
            foreach (var q in primeFactorsOfPMinus1)
            {
                if (ModPow(g, m / q, p) == BigInteger.One) { ok = false; break; }
            }
            if (ok) return g;
        }
        throw new Exception("primitive root not found");
    }

    static BigInteger[] DiscreteLogTable(BigInteger p, BigInteger root)
    {
        int size = (int)p;
        var table = new BigInteger[size];
        for (int i = 0; i < size; i++) table[i] = -1;
        BigInteger x = BigInteger.One;
        for (BigInteger k = 0; k < p - 1; k++)
        {
            table[(int)x] = k;
            x = (x * root) % p;
        }
        return table;
    }

    static BigInteger GcdBig(BigInteger a, BigInteger b) => BigInteger.GreatestCommonDivisor(a, b);

    static readonly Dictionary<int, (BigInteger, double)> S_CACHE = new Dictionary<int, (BigInteger, double)>();

    static (BigInteger, double) SForPrime(int p)
    {
        if (S_CACHE.TryGetValue(p, out var cached)) return cached;
        if (p == 2)
        {
            var res = (BigInteger.One, 0.0);
            S_CACHE[p] = res;
            return res;
        }
        BigInteger pu = new BigInteger(p);
        BigInteger m = pu - 1;
        var factors = FactorN(m);
        var divs = DivisorsFromFactorization(factors.Select(f => (f.Item1, f.Item2)).ToList());
        var primeQs = factors.Select(f => f.Item1).ToList();
        var root = PrimitiveRoot(pu, primeQs);
        var dlog = DiscreteLogTable(pu, root);

        int neededCCount = divs.Count - 1;
        var leastPrimeForC = new Dictionary<BigInteger, int>();
        foreach (int q in PRIMES)
        {
            if (q == p) continue;
            BigInteger qu = new BigInteger(q);
            int idx = (int)(qu % pu);
            if (dlog[idx] == -1) continue;
            BigInteger c = GcdBig(dlog[idx], m);
            if (c < m && !leastPrimeForC.ContainsKey(c))
            {
                leastPrimeForC[c] = q;
                if (leastPrimeForC.Count == neededCCount) break;
            }
        }
        if (leastPrimeForC.Count != neededCCount) throw new Exception("increase PRIME_SEARCH_LIMIT");

        var cItems = leastPrimeForC.Select(kv => (kv.Key, kv.Value)).ToList();

        var bestByM = new Dictionary<BigInteger, Dictionary<BigInteger, int>>();
        foreach (var mm in divs)
        {
            if (mm == BigInteger.One) continue;
            var best = new Dictionary<BigInteger, int>();
            foreach (var (c, q) in cItems)
            {
                BigInteger d = GcdBig(c, mm);
                if (d < mm)
                {
                    if (!best.ContainsKey(d) || q < best[d]) best[d] = q;
                }
            }
            bestByM[mm] = best;
        }

        var dpValue = new Dictionary<BigInteger, BigInteger>();
        var dpLog = new Dictionary<BigInteger, double>();
        dpValue[BigInteger.One] = BigInteger.One;
        dpLog[BigInteger.One] = 0.0;

        foreach (var h in divs)
        {
            if (!dpValue.ContainsKey(h)) continue;
            BigInteger mm = m / h;
            if (mm == BigInteger.One) continue;
            var best = bestByM[mm];
            BigInteger baseValue = dpValue[h];
            double baseLog = dpLog[h];
            foreach (var l in divs)
            {
                if (l > BigInteger.One && mm % l == BigInteger.Zero)
                {
                    BigInteger nextH = h * l;
                    BigInteger dKey = mm / l;
                    if (!best.ContainsKey(dKey)) continue;
                    int q = best[dKey];
                    BigInteger qu = new BigInteger(q);
                    BigInteger candidate = baseValue;
                    for (BigInteger i = 0; i < l - 1; i++) candidate *= qu;
                    double candLog = baseLog + (double)(l - 1) * Math.Log10(q);
                    if (!dpValue.ContainsKey(nextH) || candidate < dpValue[nextH])
                    {
                        dpValue[nextH] = candidate;
                        dpLog[nextH] = candLog;
                    }
                }
            }
        }

        var result = (dpValue[m], dpLog[m]);
        S_CACHE[p] = result;
        return result;
    }

    static BigInteger ProductT(int limit)
    {
        BigInteger product = BigInteger.One;
        foreach (int p in PRIMES)
        {
            if (p >= limit) break;
            product *= SForPrime(p).Item1;
        }
        return product;
    }

    static string ScientificFromInt(BigInteger n, int places = 5)
    {
        string digits = n.ToString();
        int exponent = digits.Length - 1;
        int significant = places + 1;
        string mantissaDigits;
        int finalExp;
        if (digits.Length > significant)
        {
            long head = long.Parse(digits.Substring(0, significant));
            if (int.Parse(digits[significant].ToString()) >= 5) head += 1;
            if (head == (long)Math.Pow(10, significant))
            {
                head /= 10;
                finalExp = exponent + 1;
            }
            else
            {
                finalExp = exponent;
            }
            mantissaDigits = head.ToString().PadLeft(significant, '0');
        }
        else
        {
            long headVal = long.Parse(digits);
            long head = headVal * (long)Math.Pow(10, significant - digits.Length);
            mantissaDigits = head.ToString().PadLeft(significant, '0');
            finalExp = exponent;
        }
        string mantissa = mantissaDigits[0] + "." + mantissaDigits.Substring(1);
        return $"{mantissa}e{finalExp}";
    }

    static void RunTests()
    {
        System.Diagnostics.Debug.Assert(SForPrime(2).Item1 == BigInteger.One);
        System.Diagnostics.Debug.Assert(SForPrime(5).Item1 == new BigInteger(8));
        System.Diagnostics.Debug.Assert(ProductT(20) == new BigInteger(1348422598656L));
        System.Diagnostics.Debug.Assert(ScientificFromInt(ProductT(100)) == "1.37451e123");
    }

    static void Main()
    {
        RunTests();
        Console.WriteLine(ScientificFromInt(ProductT(LIMIT)));
    }
}

File: 995/995.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/995.py

use std::collections::HashMap;

const LIMIT: usize = 20_000;
const PRIME_SEARCH_LIMIT: usize = 2_000_000;

fn sieve(n: usize) -> Vec<usize> {
    let mut is_prime = vec![true; n + 1];
    if n >= 1 { is_prime[0] = false; }
    if n >= 2 { is_prime[1] = false; }
    let r = (n as f64).sqrt() as usize;
    for i in 2..=r {
        if is_prime[i] {
            let start = i * i;
            let mut j = start;
            while j <= n {
                is_prime[j] = false;
                j += i;
            }
        }
    }
    (0..=n).filter(|&i| is_prime[i]).collect()
}

fn factor(n: u64, primes: &[usize]) -> Vec<(u64, u32)> {
    let mut out = Vec::new();
    let mut t = n;
    for &p in primes {
        let p = p as u64;
        if p * p > t {
            break;
        }
        if t % p == 0 {
            let mut e = 0u32;
            while t % p == 0 {
                t /= p;
                e += 1;
            }
            out.push((p, e));
        }
    }
    if t > 1 {
        out.push((t, 1));
    }
    out
}

fn divisors_from_factorization(factors: &[(u64, u32)]) -> Vec<u64> {
    let mut divs: Vec<u64> = vec![1];
    for &(p, e) in factors {
        let old = divs.clone();
        divs = Vec::new();
        let mut power: u64 = 1;
        for _ in 0..=(e as usize) {
            for &d in &old {
                divs.push(d * power);
            }
            power *= p;
        }
    }
    divs.sort();
    divs
}

fn primitive_root(p: u64, prime_factors_of_p_minus_1: &[u64]) -> u64 {
    if p == 2 {
        return 1;
    }
    let m = p - 1;
    for g in 2..p {
        let mut ok = true;
        for &q in prime_factors_of_p_minus_1 {
            if mod_pow(g, m / q, p) == 1 {
                ok = false;
                break;
            }
        }
        if ok {
            return g;
        }
    }
    panic!("primitive root not found");
}

fn mod_pow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = (result as u128 * base as u128 % modulus as u128) as u64;
        }
        exp >>= 1;
        base = (base as u128 * base as u128 % modulus as u128) as u64;
    }
    result
}

fn discrete_log_table(p: u64, root: u64) -> Vec<i64> {
    let mut table = vec![-1i64; p as usize];
    let mut x = 1u64;
    for k in 0..(p - 1) {
        table[x as usize] = k as i64;
        x = (x * root) % p;
    }
    table
}

fn gcd(a: u64, b: u64) -> u64 {
    if b == 0 { a } else { gcd(b, a % b) }
}

fn big_pow(base: u64, exp: u64) -> num_bigint::BigUint {
    use num_bigint::BigUint;
    use num_traits::One;
    let mut result = BigUint::one();
    let b = BigUint::from(base);
    for _ in 0..exp {
        result *= &b;
    }
    result
}

fn s_for_prime(
    p: usize,
    primes: &[usize],
    cache: &mut HashMap<usize, (num_bigint::BigUint, f64)>,
) -> (num_bigint::BigUint, f64) {
    use num_bigint::BigUint;
    use num_traits::One;

    if let Some(v) = cache.get(&p) {
        return v.clone();
    }
    if p == 2 {
        let res = (BigUint::one(), 0.0f64);
        cache.insert(p, res.clone());
        return res;
    }
    let pu = p as u64;
    let m = pu - 1;
    let factors = factor(m, primes);
    let divs = divisors_from_factorization(&factors);
    let prime_qs: Vec<u64> = factors.iter().map(|&(q, _)| q).collect();
    let root = primitive_root(pu, &prime_qs);
    let dlog = discrete_log_table(pu, root);

    let needed_c_count = divs.len() - 1;
    let mut least_prime_for_c: HashMap<u64, usize> = HashMap::new();
    for &q in primes {
        if q == p { continue; }
        let qu = q as u64;
        let idx = (qu % pu) as usize;
        if dlog[idx] < 0 { continue; }
        let c = gcd(dlog[idx] as u64, m);
        if c < m && !least_prime_for_c.contains_key(&c) {
            least_prime_for_c.insert(c, q);
            if least_prime_for_c.len() == needed_c_count {
                break;
            }
        }
    }
    if least_prime_for_c.len() != needed_c_count {
        panic!("increase PRIME_SEARCH_LIMIT");
    }

    let c_items: Vec<(u64, usize)> = least_prime_for_c.iter().map(|(&c, &q)| (c, q)).collect();

    let mut best_by_m: HashMap<u64, HashMap<u64, usize>> = HashMap::new();
    for &mm in &divs {
        if mm == 1 { continue; }
        let mut best: HashMap<u64, usize> = HashMap::new();
        for &(c, q) in &c_items {
            let d = gcd(c, mm);
            if d < mm {
                let entry = best.entry(d).or_insert(usize::MAX);
                if q < *entry {
                    *entry = q;
                }
            }
        }
        best_by_m.insert(mm, best);
    }

    let mut dp_value: HashMap<u64, BigUint> = HashMap::new();
    let mut dp_log: HashMap<u64, f64> = HashMap::new();
    dp_value.insert(1, BigUint::one());
    dp_log.insert(1, 0.0f64);

    for &h in &divs {
        if !dp_value.contains_key(&h) { continue; }
        let mm = m / h;
        if mm == 1 { continue; }
        let best = best_by_m[&mm].clone();
        let base_value = dp_value[&h].clone();
        let base_log = dp_log[&h];
        for &l in &divs {
            if l > 1 && mm % l == 0 {
                let next_h = h * l;
                let d_key = mm / l;
                if let Some(&q) = best.get(&d_key) {
                    let qu = q as u64;
                    let candidate = &base_value * big_pow(qu, l - 1);
                    let cand_log = base_log + (l - 1) as f64 * (qu as f64).log10();
                    if !dp_value.contains_key(&next_h) || candidate < dp_value[&next_h] {
                        dp_value.insert(next_h, candidate);
                        dp_log.insert(next_h, cand_log);
                    }
                }
            }
        }
    }

    let res = (dp_value[&m].clone(), dp_log[&m]);
    cache.insert(p, res.clone());
    res
}

fn product_t(
    limit: usize,
    primes: &[usize],
    cache: &mut HashMap<usize, (num_bigint::BigUint, f64)>,
) -> num_bigint::BigUint {
    use num_bigint::BigUint;
    use num_traits::One;
    let mut product = BigUint::one();
    for &p in primes {
        if p >= limit { break; }
        product *= s_for_prime(p, primes, cache).0;
    }
    product
}

fn scientific_from_int(n: &num_bigint::BigUint, places: usize) -> String {
    let digits = n.to_string();
    let exponent = digits.len() - 1;
    let significant = places + 1;
    let (mantissa_digits, final_exp) = if digits.len() > significant {
        let mut head: u64 = digits[..significant].parse().unwrap();
        let next_digit: u64 = digits[significant..significant + 1].parse().unwrap();
        if next_digit >= 5 {
            head += 1;
        }
        let sig_u = significant as u32;
        if head == 10u64.pow(sig_u) {
            head /= 10;
            (format!("{:0>width$}", head, width = significant), exponent + 1)
        } else {
            (format!("{:0>width$}", head, width = significant), exponent)
        }
    } else {
        let head_val: u64 = digits.parse().unwrap();
        let head = head_val * 10u64.pow((significant - digits.len()) as u32);
        (format!("{:0>width$}", head, width = significant), exponent)
    };
    let mantissa = format!("{}.{}", &mantissa_digits[..1], &mantissa_digits[1..]);
    format!("{}e{}", mantissa, final_exp)
}

fn run_tests(primes: &[usize], cache: &mut HashMap<usize, (num_bigint::BigUint, f64)>) {
    use num_bigint::BigUint;
    use std::str::FromStr;
    assert_eq!(s_for_prime(2, primes, cache).0, BigUint::from(1u64));
    assert_eq!(s_for_prime(5, primes, cache).0, BigUint::from(8u64));
    assert_eq!(
        product_t(20, primes, cache),
        BigUint::from_str("1348422598656").unwrap()
    );
    assert_eq!(
        scientific_from_int(&product_t(100, primes, cache), 5),
        "1.37451e123"
    );
}

fn main() {
    let primes = sieve(PRIME_SEARCH_LIMIT);
    let mut cache: HashMap<usize, (num_bigint::BigUint, f64)> = HashMap::new();
    run_tests(&primes, &mut cache);
    println!("{}", scientific_from_int(&product_t(LIMIT, &primes, &mut cache), 5));
}